src/protocol/resp/src/parse/error.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Pelikan Foundation LLC.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use std::borrow::Cow;
+use std::fmt;
+
+use bstr::BStr;
+use thiserror::Error;
+
+#[derive(Error)]
+pub enum ParseError<'a> {
+    #[error("ran out of data unexpectedly")]
+    Incomplete,
+    #[error("expected {:?}, got {:?} instead", BStr::new(.expected), BStr::new(.found))]
+    InvalidLiteral {
+        expected: &'static [u8],
+        found: Cow<'a, [u8]>,
+    },
+    #[error("expected a number, got {:?} instead", BStr::new(.0))]
+    InvalidNumber(Cow<'a, [u8]>),
+    #[error("expected a non-negative number, got {:?} instead", BStr::new(.0))]
+    UnexpectedNegativeNumber(Cow<'a, [u8]>),
+    #[error("expected a non-nil string, got a nil one instead")]
+    UnexpectedNilString,
+
+    #[error("expected an array element, but array was too short")]
+    ExpectedArrayElement,
+    #[error("expected array to be completely parsed, but there were more elements")]
+    UnexpectedArrayElement,
+
+    #[error("expected {expected}, got {:?}", BStr::new(&[*.found]))]
```

Comment:
ditto, I suggest adding quotation marks.

Reply:
In this case `expected` is a message fragment like `"+" or "$"` so it should not be quoted.